spring-integration-hazelcast/src/main/java/org/springframework/integration/hazelcast/inbound/HazelcastClusterMonitorMessageProducer.java

```diff
@@ -0,0 +1,254 @@
+					registrationId);
+		}
+
+		if (this.monitorTypes.contains(ClusterMonitorType.LIFECYCLE.name())) {
+			final String registrationId = this.hazelcastInstance.getLifecycleService()
+					.addLifecycleListener(new HazelcastLifecycleListener());
+			this.hazelcastRegisteredListenerIdMap.put(ClusterMonitorType.LIFECYCLE,
+					registrationId);
+		}
+
+		if (this.monitorTypes.contains(ClusterMonitorType.CLIENT.name())) {
+			final String registrationId = this.hazelcastInstance.getClientService()
+					.addClientListener(new HazelcastClientListener());
+			this.hazelcastRegisteredListenerIdMap.put(ClusterMonitorType.CLIENT,
+					registrationId);
+		}
+	}
+
+	@Override
+	protected void doStop() {
+		if (this.hazelcastInstance.getLifecycleService().isRunning()) {
+			if (this.monitorTypes.contains(ClusterMonitorType.MEMBERSHIP.name())) {
+				this.hazelcastInstance.getCluster().removeMembershipListener(
+						this.hazelcastRegisteredListenerIdMap
+								.get(ClusterMonitorType.MEMBERSHIP));
+				this.hazelcastRegisteredListenerIdMap
+						.remove(ClusterMonitorType.MEMBERSHIP);
+			}
+
+			if (this.monitorTypes.contains(ClusterMonitorType.DISTRIBUTED_OBJECT.name())) {
+				this.hazelcastInstance
+						.removeDistributedObjectListener(this.hazelcastRegisteredListenerIdMap
+								.get(ClusterMonitorType.DISTRIBUTED_OBJECT));
+				this.hazelcastRegisteredListenerIdMap
+						.remove(ClusterMonitorType.DISTRIBUTED_OBJECT);
+			}
+
+			if (this.monitorTypes.contains(ClusterMonitorType.MIGRATION.name())) {
+				this.hazelcastInstance.getPartitionService().removeMigrationListener(
+						this.hazelcastRegisteredListenerIdMap
+								.get(ClusterMonitorType.MIGRATION));
+				this.hazelcastRegisteredListenerIdMap
+						.remove(ClusterMonitorType.MIGRATION);
+			}
+
+			if (this.monitorTypes.contains(ClusterMonitorType.LIFECYCLE.name())) {
+				this.hazelcastInstance.getLifecycleService().removeLifecycleListener(
+						this.hazelcastRegisteredListenerIdMap
+								.get(ClusterMonitorType.LIFECYCLE));
+				this.hazelcastRegisteredListenerIdMap
+						.remove(ClusterMonitorType.LIFECYCLE);
+			}
+
+			if (this.monitorTypes.contains(ClusterMonitorType.CLIENT.name())) {
+				this.hazelcastInstance.getClientService().removeClientListener(
+						this.hazelcastRegisteredListenerIdMap
+								.get(ClusterMonitorType.CLIENT));
+				this.hazelcastRegisteredListenerIdMap.remove(ClusterMonitorType.CLIENT);
+			}
+		}
+	}
+
+	@Override
+	public String getComponentType() {
+		return "hazelcast:cm-inbound-channel-adapter";
+	}
+
+	private void processEvent(Object event) {
+		Assert.notNull(event, "'hazelcast event' must not be null");
+		this.sendMessage(getMessageBuilderFactory().withPayload(event).build());
+
+		if (logger.isDebugEnabled()) {
+			logger.debug("Received Cluster Monitor Event : " + event);
+		}
+	}
+
+	private class HazelcastMembershipListener implements MembershipListener {
```

Comment:
Looks like we can just collapse all these classes to single one with implementation for all those interfaces.
And even more - just have one `final` instance!